Make application of `nest_asyncio` patch explicit

sphuber opened the pull request:
Fixes #178 

A new event loop policy is implemented `PlumpyEventLoopPolicy` which
overrides the `get_event_loop` method to return an event loop that is
patched with the `nest_asyncio` library to make it reentrant. It also
makes sure that it always returns the same event loop because once made
reentrant, only a single one is required.

The policy now has to be enabled explicitly by calling the method
`set_event_loop_policy` instead of implicit, as it used to be done in
the constructor of the `Process` class. The event loop policy can be
reset through `reset_event_loop_policy`.

muhrin commented on plumpy/events.py:
```diff
@@ -3,13 +3,42 @@
+
+    _loop = None
+
+    def get_event_loop(self):
```
I'm not sure if we'll overwrite `new_event_loop` to do the same thing as this.  I'm thinking about the case of a child runner where, I guess, it calls `new_event_loop` when the runner gets instantiated.  We could, of course, change the runner code but it feels like whether the loop is reentrant or not should be transparent to clients (at least as much as possible).  One problem with this is that if a client calls `new_event_loop` there is no way for us to know if they were just planning to run a new loop in serial after destroying the old one, or to, as in AiiDA, they were planning to run the loop inside the current one...

sphuber replied:
I am not sure either. For what it's worth, neither in `plumpy`'s nor in `aiida-core`'s source code, is `new_event_loop` ever called. That doesn't help with the problem for other users of this library of course. Though I am not sure if `plumpy` is even being used outside of `aiida-core`.

sphuber replied:
@muhrin Coming back to this, because ideally we need to finish this up a.s.a.p. If the current loop is destroyed before they call `new_event_loop` they will get a new reentrant loop, even if we override it to behave exactly as our `get_event_loop`, wouldn't it? Because if they delete it, the global loop that `get_event_loop` would return also no longer exists and so a new one is created. The problem, however, might be what happens when they call `new_event_loop` without having deleted the current one, if we override `new_event_loop` to be the same as `get_event_loop`, then that behavior might be surprising. Maybe we should just document that `plumpy` works this way though and only ever has a single reentrant loop. Maybe even explicitly raise `NotImplementedError` for `new_event_loop`? Anyway, since this is also not relevant for `aiida-core` for now, and `plumpy` does not really have other users, maybe we should make a choice and release it. If we later on find out it is not ideal for a concrete reason we can fix it.

unkcpz replied:
> Maybe even explicitly raise `NotImplementedError` for `new_event_loop`?

I vote for this. We do not need to worry about the `new_event_loop` behaviour if it not be used anywhere else at the moment, isn't it? And raise `NotImplementedError` to avoid strange behaviour from happening in the future.

muhrin replied:
Sorry for the delay - there have been a lot of meeting recently.

OK, so my suggestion about what `new_event_loop` should do was based on a misunderstanding on my part on what `asyncio` does natively. I've just rechecked the code and actually, `new_event_loop` should probably just return a new instance of a reentrant event loop.  But I'm also fine with the `NotImplementedError` given that it's not used.

sphuber replied:
If the `new_event_loop` normally does simply destroy the global loop and then create a new one, we could simply replicate this, couldn't we? What happens with existing references to the original loop though? Is this is also what happens with vanilla `asyncio`? Will `new_event_loop` actually destroy some existing event loop instances? I thought the idea was there that you could have multiple event loops but just not have them nested?

muhrin replied:
Well, [normally](https://github.com/python/cpython/blob/master/Lib/asyncio/events.py#L653), `new_event_loop` does nothing more than instantiate a new loop.  Doesn't destroy anything or set that new loop to be the current loop so if we do the same then we're fulfilling the contract.  And, yes, you're right you can have multiple, non-nested, loops (which should also work with multiple instance of reentrant loops).

muhrin reviewed (commented):
Code is good - just wanted to get your thoughts on what comment I made before we move on.

sphuber commented:
@muhrin one additional comment here is that I haven't added any code to undo the patching of the event loop by `nest_asyncio` other than just setting the default event loop policy. @unkcpz included code in his PR to do this explicitly. He suggested that this should rather go into `nest_asyncio` itself as some kind of `reset` method, and I agree with him that it should go there. Should we already include some behavior like this here, whether directly in `plumpy` or by first submitting a patch to `nest_asyncio`?

muhrin commented:
Yes, the reset should be in `nest_asyncio` (given that `nest_asyncio` knows what it changed).  I guess it'll take time to get it in there I think we should just implement `reset()` ourselves and then just change what we call once/if it gets merged.  

Then a separate question is one of _when_ to call the reset().  It's not 100% clear but given that we're basically hijacking the loop class it'd be good to make at least some effort to have a clean up if the user or tests need this behaviour.  So I guess in the `reset...policy` function this can be done but we might need to keep a track of which loop classes we have changed in the policy itself.

sphuber commented:
> So I guess in the `reset...policy` function this can be done but we might need to keep a track of which loop classes we have changed in the policy itself.

I am not sure I understand. How can we keep track of classes we are modifying if we are not actually modifying any? If `nest_asyncio` is doing this, then their `reset` method should take care of this. If they do the correct thing, us calling their `reset` in our `reset_event_loop_pollicy` should be sufficient, shouldn't it?

muhrin commented:
Well the interface is a little fuzzy here but say you're actually passing the loop to `nest_asyncio.apply()` as we do.  Then even if there were a `reset(loop)` call in `nest_asyncio` it would be up to us to 'remember' which loops we called `apply` on.  All this, while keeping in mind that, in fact, `apply` [doesn't really operate on](https://github.com/erdewit/nest_asyncio/blob/master/nest_asyncio.py#L139) a loop instance, but rather on the class and so this is really the important thing to track.

sphuber commented:
OK, @muhrin let's keep `new_event_loop` as `NotImplemented` for now. If we find in the future that we need it and can provide a correct and consistent implementation, we can add it. I have updated the `reset_event_loop_policy` to include the reset code for `nest_asyncio` from @unkcpz . I have added him as contributor to the commit. Please review this PR and see if you are happy with it. I will now test it with the aiida-core branch and make sure things seem to work.

sphuber commented:
I have run my production test with this branch and it seems to work fine.

muhrin reviewed (approved):
Ok, I think is is a fine compromise.
